Mathlib/CategoryTheory/Sites/RegularExtensive.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson, Filippo A. E. Nuccio, Riccardo Brasca
+-/
+import Mathlib.CategoryTheory.Sites.Coherent
+/-!
+
+# The Regular and Extensive Coverages
+
+This file defines two coverages on a category `C`.
+
+The first one is called the *regular* coverage and for that to exist, the category `C` must satisfy
+a condition called `Regular C`. This means that effective epimorphisms can be "pulled back". The
+covering sieves of this coverage are generated by presieves consisting of a single effective
+epimorphism.
+
+The second one is called the *extensive* coverage and for that to exist, the category `C` must
+satisfy a condition called `Extensive C`. This means `C` has finite coproducts and that those
+are preserved by pullbacks. The covering sieves of this coverage are generated by presieves
+consisting finitely many arrows that together induce an isomorphism from the coproduct to the
+target.
+
+In `extensive_union_regular_generates_coherent`, we prove that the union of these two coverages
+generates the coherent topology on `C` if `C` is precoherent, extensive and regular.
+
+TODO: figure out under what conditions `Regular` and `Extensive` are implied by `Precoherent` and
+vice versa.
+
+-/
+
+universe v u w
+
+namespace CategoryTheory
+
+open Limits
+
+variable (C : Type u) [Category.{v} C]
+
+/--
+The condition `Regular C` is property that effective epis can be "pulled back" along any
+morphism. This is satisfied e.g. by categories that have pullbacks that preserve effective
+epimorphisms (like `Profinite` and `CompHaus`), and categories where every object is projective
+(like  `Stonean`).
+-/
+class Regular : Prop where
```

Comment:
This is not what it is usually called a regular category, right? I mean, like [here](https://ncatlab.org/nlab/show/regular+category).

Reply:
It's usually required to have finite limits, yes. But this is all that's needed to define the regular coverage. Should we go with `Preregular`?

Reply:
`Preregular` seems a fine name. The coverage can still be called `Regular` I think.